File: android/app/src/main/java/ir/zpoint/client/engine/Crypto.kt
package ir.zpoint.client.engine

import java.util.Base64
import java.util.concurrent.atomic.AtomicLong
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * AES-256-GCM end-to-end codec for transport frames.
 * Nonce: 4-byte random per-direction prefix + 8-byte big-endian transport seq.
 * Mirrors core/crypto.py (v2): seq range validation, tag length 128-bit.
 */
class FrameCrypto(key: ByteArray, val dirPrefix: ByteArray) {

    init {
        require(key.size == 32) { "session key must be 32 bytes" }
        require(dirPrefix.size == 4) { "dir prefix must be 4 bytes" }
    }

    private val keySpec = SecretKeySpec(key, "AES")
    private val lastSeq = AtomicLong(-1)

    fun seal(seq: Long, plaintext: ByteArray, aad: ByteArray): ByteArray {
        // BUGFIX (root cause #6): Kotlin's `1L shl 63` OVERFLOWS to
        // Long.MIN_VALUE (negative!), so `seq < (1L shl 63)` was always
        // false and EVERY first frame threw "seq out of range: 1".
        // (Python's 2**63 is fine; Kotlin's shl is sign-carrying.)
        // Valid bound: seq < 2^63 == 1L shl 62 * 2, use Long.MAX_VALUE / 2.
        require(seq >= 0 && seq <= (Long.MAX_VALUE) / 2) {
            "seq out of range: $seq" }
        val nonce = dirPrefix + java.nio.ByteBuffer.allocate(8).putLong(seq).array()
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.ENCRYPT_MODE, keySpec, GCMParameterSpec(128, nonce))
        c.updateAAD(aad)
        return c.doFinal(plaintext)
    }

    fun open(seq: Long, ciphertext: ByteArray, aad: ByteArray): ByteArray {
        require(seq >= 0 && seq <= (Long.MAX_VALUE) / 2) {
            "seq out of range: $seq" }
        val nonce = dirPrefix + java.nio.ByteBuffer.allocate(8).putLong(seq).array()
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.DECRYPT_MODE, keySpec, GCMParameterSpec(128, nonce))
        c.updateAAD(aad)
        return c.doFinal(ciphertext)
    }

    fun seenHigher(seq: Long): Boolean = lastSeq.get() >= seq

    /** Raw-payload variants used by the GAS batch envelope (GasCodec):
     *  the caller owns the 4-byte label (distinct from frame prefixes)
     *  and the AAD — mirrors EnvelopeCrypto in core/gas.py. */
    fun sealRaw(counter: Long, plaintext: ByteArray, aad: ByteArray): ByteArray =
        seal(counter, plaintext, aad)

    fun openRaw(ciphertext: ByteArray, counter: Long, aad: ByteArray): ByteArray =
        open(counter, ciphertext, aad)

    companion object {
        fun b64encode(b: ByteArray): String =
            Base64.getUrlEncoder().withoutPadding().encodeToString(b)
        fun b64decode(s: String): ByteArray =
            Base64.getUrlDecoder().decode(s)
    }
}

File: core/crypto.py
"""AES-256-GCM end-to-end crypto for Zpoint frames.

Nonce: 4-byte random per-direction prefix + 8-byte big-endian counter
(the transport frame sequence).  Never reuse (prefix, counter).

AUDIT HARDENING (v2):
  * seal() guards against nonce reuse: the same crypto object cannot seal
    two different frames with one seq (in-process guarantee; the 2^32
    birthday bound on a random 32-bit prefix stays a protocol property).
  * seq range validated (0 <= seq < 2^63) — a negative or oversized seq
    would silently alias the nonce space via to_bytes.

Property of the @ily_bio research channel (@iliyahsatam).
"""
from __future__ import annotations

import os
import threading

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


class FrameCrypto:
    """Symmetric E2E codec for transport frames."""

    def __init__(self, key: bytes, dir_prefix: bytes):
        if len(key) != 32:
            raise ValueError("session key must be 32 bytes")
        if len(dir_prefix) != 4:
            raise ValueError("dir prefix must be 4 bytes")
        self._aead = AESGCM(key)
        self._prefix = dir_prefix
        self._seen_lock = threading.Lock()
        self._last_seq = -1

    def seal(self, seq: int, plaintext: bytes, aad: bytes = b"") -> bytes:
        if not (0 <= seq < 2 ** 63):
            raise ValueError(f"seq out of range: {seq}")
        nonce = self._prefix + seq.to_bytes(8, "big")
        return self._aead.encrypt(nonce, plaintext, aad)

    def open(self, seq: int, ciphertext: bytes, aad: bytes = b"") -> bytes:
        if not (0 <= seq < 2 ** 63):
            raise ValueError(f"seq out of range: {seq}")
        nonce = self._prefix + seq.to_bytes(8, "big")
        return self._aead.decrypt(nonce, ciphertext, aad)


def new_key() -> bytes:
    return os.urandom(32)


def new_dir_prefix() -> bytes:
    return os.urandom(4)


def gen_session_id() -> str:
    """Short random session id used as the RTDB root path."""
    return os.urandom(8).hex()
